ompi/mca/part/persist/part_persist.c

```diff
@@ -39,6 +47,492 @@ ompi_part_persist_t ompi_part_persist = {
     }
 };
 
+/**
+ * This is a helper function that frees a request. This requires ompi_part_persist.lock be held before calling.
+ */
+__opal_attribute_always_inline__ static inline int
+mca_part_persist_free_req(struct mca_part_persist_request_t* req)
+{
+    int err = OMPI_SUCCESS;
+    size_t i;
+    opal_list_remove_item(ompi_part_persist.progress_list, (opal_list_item_t*)req->progress_elem);
+    OBJ_RELEASE(req->progress_elem);
+
+    for(i = 0; i < req->real_parts; i++) {
+        ompi_request_free(&(req->persist_reqs[i]));
+    }
+    free(req->persist_reqs);
+    free(req->flags);
+
+    if( MCA_PART_PERSIST_REQUEST_PRECV == req->req_type ) {
+        MCA_PART_PERSIST_PRECV_REQUEST_RETURN(req);
+    } else {
+        MCA_PART_PERSIST_PSEND_REQUEST_RETURN(req);
+    }
+    return err;
+}
+
+static void
+mca_part_persist_complete(struct mca_part_persist_request_t* request)
+{
+    if(MCA_PART_PERSIST_REQUEST_PRECV == request->req_type) {
+        request->req_ompi.req_status.MPI_SOURCE = request->req_peer;
+    } else {
+        request->req_ompi.req_status.MPI_SOURCE = request->req_comm->c_my_rank;
+    }
+    request->req_ompi.req_complete_cb = NULL;
```

Comment:
This is not the correct behavior for requests. Setting `req_complete_cb` to NULL here will prevent the `ompi_request_complete` from properly calling the callback associated with the request. This means that some of the collectives components that use this functionality could not use partitioned communications (I'm not saying they do, just that this breakage of consistency will be a pain to debug in the future).